### ISSUE TITLE
Bluesky new logging framework

### ISSUE THREAD

**Issue by ke-zhang-rd** [bluesky/bluesky#1185]
## Expected Behavior

Control flow on handler level 
## Current Behavior

Control flow on logger level
## Possible Solution

new log.py file copy paster most from caproto log [here](https://github.com/caproto/caproto/blob/master/caproto/_log.py) 
## Steps to Reproduce (for bugs)

1.
2.
3.

**Comment by danielballan** [comment]
## Important use cases we want to support

* Log each `Msg` before it is processed.
* When a document is emitted to the callback registry, log its type and uid.
* Log which status objects we are waiting on.
* Log when status objects finish.

# List of loggers we need
* `bluesky`
* `bluesky.emit_document`
* `bluesky.msg`
* `bluesky.status`

## List of extras we need (where applicable)

* `uid`
* `repr(msg)`

**Comment by ke-zhang-rd** [comment]
What do you think could be tags for bluesky log, or do we need them @danielballan

**Comment by danielballan** [comment]
I listed the tags I had in mind under "List of extras" above. Seems like we have similar ideas. What do you mean by "old_status" vs "new_status"? Are you referring to `RE.state`?

**Comment by danielballan** [comment]
For anyone else following along, @ke-zhang-rd has updated the comment at the top of this issue in place to reflect our current plan.

**Comment by danielballan** [comment]
I think the idea of labeling the msg-related log lines is good, but I would prefer less SHOUT-y capitalization and `[]`s. Something like:

```
[D 17:46:32.405      run_engine: 1188] msg save: (None), (), {}
```

**Comment by danielballan** [comment]
On the train yesterday, @tacaswell and @ke-zhang-rd went back and forth about whether a log record should be generated and, if so, how much information should be in that record. Proposals:

1. Do not log the documents at all. We have `RE.subscribe` for that.
2. Log the whole document and let the handler/formatter decide how much of it to incorporate.
3. Log just the document type.
4. Log the document type and the uid of the current Run.

One rejected possibility was to log the document type and document uid. This is not tenable for EventPages and DatumPages, which have a column of uids, not one uid.

Currently, I have a slight preference is (4) but I think this needs more thought and discussion.

**Comment by ke-zhang-rd** [comment]
> On the train yesterday, @tacaswell and @ke-zhang-rd and I went back and forth about whether a log record should be generated and, if so, how much information should be in that record. Proposals:
> 
> 1. Do not log the documents at all. We have `RE.subscribe` for that.
> 2. Log the whole document and let the handler/formatter decide how much of it to incorporate.
> 3. Log just the document type.
> 4. Log the document type and the uid of the current Run.
> 
> One rejected possibility was to log the document type and document uid. This is not tenable for EventPages and DatumPages, which have a column of uids, not one uid.
> 
> Currently, I have a slight preference is (4) but I think this needs more thought and discussion.

Is the uid of current Run identical for all document emit log of same Run? If yes, I will vote for 3 and include uid of Run to other config/misc log message, for instance 
```
[I 10:30:00.601      run_engine:  747] Executing plan <generator object count at 0x11b65e620>
```
 Or 
```
[I 10:30:00.602      run_engine:   86] Change state on <bluesky.run_engine.RunEngine object at 0x11b662400> from 'idle' -> 'running'
```

**Comment by danielballan** [comment]
> Is the uid of current Run identical for all document emit log of same Run?

Can you rephrase the question? I'm not sure I understand.

**Comment by ke-zhang-rd** [comment]
> > Is the uid of current Run identical for all document emit log of same Run?
> 
> Can you rephrase the question? I'm not sure I understand.

I think `self._run_start_uid` solved my problem.

**Comment by ke-zhang-rd** [comment]
For reflecting @danielballan 's summary, current log output looks like,
```
[I 11:57:15.624      run_engine:  747] Executing plan <generator object count at 0x1206a8a98>
[I 11:57:15.627      run_engine:   86] Change state on <bluesky.run_engine.RunEngine object at 0x1206b5160> from 'idle' -> 'running'
[D 11:57:15.627      run_engine: 1188] msg stage: (SynGauss(name='det1', value=5.0, timestamp=1559145435.3698492)), (), {}
[D 11:57:15.627      run_engine: 1188] msg stage: (SynGauss(name='det2', value=1.764993805169191, timestamp=1559145435.3699698)), (), {}
[D 11:57:15.627      run_engine: 1188] msg open_run: (None), (), {'detectors': ['det1', 'det2'], 'num_points': 1, 'num_intervals': 0, 'plan_args': {'detectors': ["SynGauss(name='det1', value=5.0, timestamp=1559145435.3698492)", "SynGauss(name='det2', value=1.764993805169191, timestamp=1559145435.3699698)"], 'num': 1}, 'plan_name': 'count', 'hints': {'dimensions': [(('time',), 'primary')]}}
[D 11:57:15.631      run_engine: 1380] [start] Emitted RunStart (uid='bb94771c-4dfa-414a-8545-c01171a9499f')
[D 11:57:15.633      run_engine: 1188] msg checkpoint: (None), (), {}
[D 11:57:15.633      run_engine: 1188] msg trigger: (SynGauss(name='det1', value=5.0, timestamp=1559145435.3698492)), (), {'group': 'trigger-ca98b9'}
[D 11:57:15.633      run_engine: 2013] The object SynGauss(name='det1', value=5.0, timestamp=1559145435.633526) reports trigger is done with status True.
[D 11:57:15.633      run_engine: 1188] msg trigger: (SynGauss(name='det2', value=1.764993805169191, timestamp=1559145435.3699698)), (), {'group': 'trigger-ca98b9'}
[D 11:57:15.634      run_engine: 2013] The object SynGauss(name='det2', value=1.764993805169191, timestamp=1559145435.634005) reports trigger is done with status True.
[D 11:57:15.634      run_engine: 1188] msg wait: (None), (), {'group': 'trigger-ca98b9'}
[D 11:57:15.634      run_engine: 1188] msg create: (None), (), {'name': 'primary'}
[D 11:57:15.634      run_engine: 1188] msg read: (SynGauss(name='det1', value=5.0, timestamp=1559145435.633526)), (), {}
[D 11:57:15.635      run_engine: 1188] msg read: (SynGauss(name='det2', value=1.764993805169191, timestamp=1559145435.634005)), (), {}
[D 11:57:15.636      run_engine: 1188] msg save: (None), (), {}
[D 11:57:15.638      run_engine: 1704] [descriptor] Emitted Event Descriptor with name 'primary' containing data keys dict_keys(['det1', 'det2']) (uid='9bc1899c-8a52-4bb8-875d-963a5ff98954')
[D 11:57:15.639      run_engine: 1736] [event] Emitted Event with data keys dict_keys(['det1', 'det2']) (currnt_run_uid='bb94771c-4dfa-414a-8545-c01171a9499f')
[D 11:57:15.639      run_engine: 1188] msg close_run: (None), (), {'exit_status': None, 'reason': None}
[D 11:57:15.639      run_engine: 1413] Stopping run 'bb94771c-4dfa-414a-8545-c01171a9499f'
[D 11:57:15.640      run_engine: 1439] [stop] Emitted RunStop (uid='600bbb8e-98f5-4677-92bd-08889f770f7d')
[D 11:57:15.640      run_engine: 1188] msg unstage: (SynGauss(name='det2', value=1.764993805169191, timestamp=1559145435.634005)), (), {}
[D 11:57:15.640      run_engine: 1188] msg unstage: (SynGauss(name='det1', value=5.0, timestamp=1559145435.633526)), (), {}
[I 11:57:15.641      run_engine:   86] Change state on <bluesky.run_engine.RunEngine object at 0x1206b5160> from 'running' -> 'idle'
[I 11:57:15.641      run_engine: 1319] Cleaned up from plan <generator object count at 0x1206a8a98>
```
Any comments and concerns?

**Comment by prjemian** [comment]
Wow!  That looks clear to me.

**Comment by klauer** [comment]
@ke-zhang-rd Looks pretty good! 

A couple random thoughts about messages:

What if the messages had function-call like formatting?

Currently:
```
[D 11:57:15.640      run_engine: 1188] msg unstage: (SynGauss(name='det1', value=5.0, timestamp=1559145435.633526)), (), {}
```
might be simply formatted as:
```
[D 11:57:15.640      run_engine: 1188] msg unstage(SynGauss(name='det1', value=5.0, timestamp=1559145435.633526), *(), **{})
```

Or perhaps the object repr could be pushed to the end:
```
[D 11:57:15.640      run_engine: 1188] msg det1.unstage(*(), **{}) where det1=SynGauss(name='det1', value=5.0, timestamp=1559145435.633526)
```

Overall, I think the reprs get in the way of easy readability. Not a big concern, though.

Failing that, I'd just recommend cutting back on unnecessary punctuation. It may seem trivial on this scale of log output, but it really adds up when searching through a large log dump.

**Comment by danielballan** [comment]
This seems to include the `uid` of each individual document rather than `uid` of the overall run. Is that intentional? What will happen when, in the future, an `EventPage` is emitted, which has not single unique ID? With this in mind, I suggested (4).

:+1: for @klauer's formatting suggestions

I'm also not sure that the "Emitted" log messages need to display the document name twice.

**Comment by ke-zhang-rd** [comment]
@klauer I'll use your recommendation format below since I want make sure Msg's `command` right after `msg`
```
[D 11:57:15.640      run_engine: 1188] msg unstage(SynGauss(name='det1', value=5.0, timestamp=1559145435.633526), *(), **{})
```
@danielballan 
```
[D 11:57:15.631      run_engine: 1380] [start] Emitted RunStart (uid='bb94771c-4dfa-414a-8545-c01171a9499f')
...
[D 11:57:15.639      run_engine: 1736] [event] Emitted Event with data keys dict_keys(['det1', 'det2']) (currnt_run_uid='bb94771c-4dfa-414a-8545-c01171a9499f')`
```

The Event uses uid of the overall run. I didn't make Descriptor and Stop use overall run uid. But I'll update them later.
> I'm also not sure that the "Emitted" log messages need to display the document name twice.

I think put document name first may make them cleaner, so I may rephrase words to eliminate second one.

**Comment by danielballan** [comment]
> I think put document name first may make them cleaner, so I may rephrase words to eliminate second one.

I agree. I think that's a good way to go

> The Event uses uid of the overall run. I didn't make Descriptor and Stop use overall run uid. But I'll update them later.

Gotcha.

**Comment by danielballan** [comment]
Done!
